### DISCUSSION TITLE
Subscribe resolver access to websocket

### DISCUSSION THREAD

**Discussion by jmc420** [enisdenjo/graphql-ws#465] [Q&A] [answered]
Within the subscribe method of a resolver, is it possible to find out which websocket initiated the subscription?

The reason I ask is because of the use of generator functions and async iterates in the example:

```
subscribe: async function* () {
          for (const hi of ['Hi', 'Bonjour', 'Hola', 'Ciao', 'Zdravo']) {
            yield { greetings: hi };
          }
        },
```

This makes good sense because the iterable set is finite. In a real world scenario the end of the set is typically defined by socket closure hence the request as to whether there is a way I can find out the id of the web socket is in the subscribe function.

The other alternative is to use a pubsub iterator following the Apollo server approach:

```
subscribe: () => {
            return pubsub.asyncIterator('NEW_ORDER')
        }
```

I find Apollo Server rather huge and indigestible and though there are various npm pubsub packages, none of them fit the bill. If there was access to the web socket in the subscribe function (via a context?), the publishing could be done quite simply with the nodejs eventEmitter.

**Comment by enisdenjo** [comment, marked as answer]
> If there was access to the web socket in the subscribe function (via a context?)

Exactly, this is the way to do it. You can find the [socket in the context extra](https://github.com/enisdenjo/graphql-ws/blob/f4320f3/docs/interfaces/use_ws.Extra.md#socket) which you can pass to the GraphQL context:

```js
import { WebSocketServer } from 'ws';
import { useServer } from 'graphql-ws/lib/use/ws';
import { schema } from './my-graphql-schema';

const server = new WebSocketServer({
  port: 4000,
  path: '/graphql',
});

useServer(
  {
    schema,
    // pass the socket to the GraphQL context
    context: (ctx) => ({ socket: ctx.extra.socket }),
  },
  server,
);
```

**Reply by jmc420** [reply]
thanks
